app/ante/ante.go

```diff
@@ -94,28 +98,55 @@ func NewAnteHandler(options ethante.HandlerOptions) (sdk.AnteHandler, error) {
 		// handle as totally normal Cosmos SDK tx
 		switch tx.(type) {
 		case sdk.Tx:
-			found := false
-			for _, msg := range tx.GetMsgs() {
-				switch msg.(type) {
-				// treat these three msg types differently because they might call EVM which results in massive gas consumption
-				// For these two msg types, we don't check gas limit by using a different ante handler
-				case *cctxtypes.MsgGasPriceVoter, *cctxtypes.MsgVoteOnObservedInboundTx:
-					found = true
-					break
-				case *stakingtypes.MsgCreateValidator:
-					if ctx.BlockHeight() == 0 {
-						found = true
-						break
-					}
+			anteHandler = newCosmosAnteHandler(options)
+			if len(tx.GetMsgs()) != 1 {
+				break
+			}
+
+			msg := tx.GetMsgs()[0] // now we must have len(tx.GetMsgs()) == 1
+			var innerMsg sdk.Msg
+			innerMsg = msg
+			if mm, ok := msg.(*authz.MsgExec); ok { // authz tx; look inside it
+				msgs, err := mm.GetMessages()
+				if err == nil && len(msgs) == 1 {
+					innerMsg = msgs[0]
 				}
 			}
-			if len(tx.GetMsgs()) == 1 && found {
-				// this differs newCosmosAnteHandler only in that it doesn't check gas limit
-				// by using an Infinite Gas Meter.
-				anteHandler = newCosmosAnteHandlerNoGasLimit(options)
-			} else {
-				anteHandler = newCosmosAnteHandler(options)
+
+			isAuthorize := options.ObserverKeeper.IsAuthorized
+			if mm, ok := innerMsg.(*cctxtypes.MsgGasPriceVoter); ok && isAuthorize(ctx, mm.Creator) {
```

Comment:
use the NoFee antehandler only for system txs that actually come from observers.